### PR TITLE
Add Deployment type 2

### LINKED ISSUE #2508: Allow users to connect an existing DB to Mathesar during installation
## Deployment Type 2

This deployment has the web service and its Django model tables DB running in Docker containers managed by `docker-compose`, but the user DB is assumed to be running separately either on the same underlying machine, or a different server altogether. Example: running the `mathesar_service` and a DB container for Django model tables on a minimal EC2 instance, and having user tables on an RDS PostgreSQL instance. N.B., we can't easily script setting up a preexisting DB with an appropriate user.

### Installation User Flow

As for Deployment Type 1, they should find the installation type they want on the website, and copy-and-paste an appropriate terminal command. The differences for the user flow are that they'll need to have installed PostgreSQL (somewhere) and figure out the needed credentials. We'll provide documentation on the associated installation page on `docs.mathesar.org` about what permissions are needed on their DB, and how to determine the credentials. The script run by the command they copy-and-paste should be extended to do the following:

1. Check the `docker` and `docker-compose` versions, and exit non-zero (with a useful message) if they're not going to work.
1. Prompt the user with a question about whether they want to install on a Mathesar-managed DB in the `docker-compose` network, or connect to a preexisting DB, and branch there. (for the rest of this user flow, assume they chose preexisting DB)
   - This step should explain what will be modified on the preexisting DB.
   - Explain the implications of the needed `mathesar` superuser in their preexisting DB.
1. Ask for an installation directory to hold config files (or we could just put it under `$HOME/.config/mathesar` as is tradition)
1. Download appropriate versions of the `docker-compose.yml` and `.env.example` from the repo to the local machine.
1. Move `.env.example` to `.env` and customize it with appropriate secrets and other variables.
   - This section would be interactive so that they can input things like DB keys, etc.
   - Prompt them appropriately to give credentials for their preexisting DB.
1. Run `docker-compose up <appropriate_service_name>`.
   - This would pull all needed images and start all necessary containers, including a Mathesar-managed DB container (for the Django DB).
1. Run `install.py` in the running container.
   - `install.py` should accept a flag to first check the DB server version of the user DB.
1. Run migrations (probably as part of `install.py`)
1. Set up a superuser
1. Output information about configuration, and how to connect a browser to Mathesar.

### Installation Flow Tasks

In addition to the pieces of type 1 above, we should need:

- [ ] Modify `install.sh` to prompt for the preexisting DB credentials, and provide the explanations / info described in the user flow.
- [ ] Modify `install.py` to accept / use those credentials as needed.
- [ ] Document use of this setup on `docs.mathesar.org` @mathemancer
- [ ] #2048

## Notes
- This was moved from https://github.com/centerofci/mathesar/issues/2034.

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #2508 

This adds functionality to `install.sh` that allows a user to set up a connection to an already-existing database.

**Technical details**

This also cleans up some parts of the code, and reduces some repetition by extracting some methods.

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `master` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: docs/docs/installation/docker-compose.md [hunk 1]
```diff
@@ -55,6 +55,10 @@ Mathesar actually uses two databases; one internal system database for metadata
 
 The credentials created in this section are used to login directly to the database (i.e., not through the Mathesar UI).
 
+**Note on permissions**: If you choose to connect Mathesar to a preexisting PostgreSQL database, you need to make sure that the user you choose:
```
Comment (kgodey) [inline docs/docs/installation/docker-compose.md:58]: I'd like to make additional documentation updates, but I can handle those myself tomorrow.

File: install.sh [hunk 2]
```diff
@@ -87,33 +190,22 @@ can be used to login directly using psql or another client.
 --------------------------------------------------------------------------------
 
 "
-read -r -p "Choose a database name [mathesar]: " db_name
-db_name=${db_name:-mathesar}
 
-read -r -p "Choose a username [mathesar]: " db_username
-db_username=${db_username:-mathesar}
+existing=$(standardize_yesno "Do you have a preexisting PostgreSQL database to connect?")
```
Comment (kgodey) [inline install.sh:194]: I would say something like "Would you like to connect an existing database or create a new database?" here. It's more clear.

File: install.sh [hunk 3]
```diff
@@ -87,33 +190,22 @@ can be used to login directly using psql or another client.
...
-  done
-  printf "\n"
-  read -rs -p "Repeat the password: " db_password_check
-done
-printf "\n"
-read -r -p "Choose a port for local database access [5432]: " db_port
-db_port=${db_port:-5432}
+if [ "${existing}" == Y ]; then
+  printf "
+WARNING: This will add a schema to the database for Mathesar functions!
```
Comment (kgodey) [inline install.sh:198]: "add a schema" --> "create a new PostgreSQL schema"

**Review by kgodey — COMMENTED** [review]
Added some suggestions for improving clarity.

File: install.sh (hunk 2 above)
Comment (mathemancer) [reply install.sh:194]: I'm not sure how to interpret the result. I wanted a "yes-no" question

File: install.sh (hunk 3 above)
Comment (mathemancer) [reply install.sh:198]: Sure, makes sense.

**Review by silentninja — CHANGES_REQUESTED** [review]
The primary issue I found was:
I am not able to connect to a database running on the host. This is because docker considers its container as the `localhost`, not the host the containers are running on.

We should document the correct `IP` address to use if we want to connect to the host database from the containers.

File: install.sh (hunk 2 above)
Comment (mathemancer) [reply install.sh]: Okay, I changed this to use a select menu.

**Comment by mathemancer** [comment]
> The primary issue I found was: I am not able to connect to a database running on the host. This is because docker considers its container as the `localhost`, not the host the containers are running on.
> 
> We should document the correct `IP` address to use if we want to connect to the host database from the containers.

I'm solving this by putting a note in the docs that we're not supporting this functionality at the moment.

**Review by silentninja — APPROVED** [review]

**Review by kgodey — APPROVED** [review]
I'm going to merge this. We can make new issues for anything we encounter during testing.

**Pull request merged**